### PR TITLE
Add configuration option for matching users with CiviCRM contacts on login

### PR DESCRIPTION
**Pull request by jensschuppe**
This adds a configuration option for matching users with CiviCRM contacts when they login:

* when matching is generally enabled
* only when they don't have a CiviRemote ID already
* when they don't have any of the configured roles to exclude

*systopia-reference: 22801*

### CODE REVIEW

File: config/install/civiremote.settings.yml [hunk 1]
```diff
@@ -1,3 +1,4 @@
 cmrf_connector: civiremote
 acquire_civiremote_id: false
 match_blocked_users: false
+match_on_login: false
```
Comment (dontub) [inline config/install/civiremote.settings.yml:4]: On existing installations the initial value should be `true` to keep the current behavior.

File: src/Form/CiviRemoteConfigForm.php [hunk 2]
```diff
@@ -16,27 +16,35 @@
...
+use Drupal\user\RoleInterface;
 use Symfony\Component\DependencyInjection\ContainerInterface;
 use Drupal\cmrf_core;
+use Drupal\user\RoleStorageInterface;
 
 class CiviRemoteConfigForm extends ConfigFormBase {
 
-  /* @var cmrf_core\Core $cmrf_core */
+  /** @var cmrf_core\Core $cmrf_core */
   public $cmrf_core;
```
Comment (dontub) [inline src/Form/CiviRemoteConfigForm.php]: PHP type hint?

File: src/Form/CiviRemoteConfigForm.php [hunk 3]
```diff
@@ -16,27 +16,35 @@
...
 
+  /** @var \Drupal\user\RoleStorageInterface $roleStorage */
+  public $roleStorage;
```
Comment (dontub) [inline src/Form/CiviRemoteConfigForm.php]: PHP type hint?

File: src/Form/CiviRemoteConfigForm.php [hunk 4]
```diff
@@ -101,6 +117,30 @@ public function buildForm(array $form, FormStateInterface $form_state) {
...
+      '#default_value' => $config->get('match_on_login'),
+      '#states' => [
+        'visible' => [':input[name="acquire_civiremote_id"]' => ['checked' => TRUE]],
+      ],
+    ];
+
+    $form['match_contacts']['match_on_login_exclude_roles'] = [
+      '#type' => 'checkboxes',
+      '#title' => $this->t('Roles to exclude when matching on login'),
+      '#description' => $this->t('Users with one of the selected roles will be escluded from acquiring a CiviCRM Contact ID during login.'),
```
Comment (dontub) [inline src/Form/CiviRemoteConfigForm.php]: s/escluded/excluded

File: config/schema/civiremote.schema.yml [hunk 5]
```diff
@@ -10,6 +10,15 @@ civiremote.settings:
     match_blocked_users:
       type: boolean
       label: 'Match blocked users'
+    match_on_login:
+      type: boolean
+      label: 'Acquire CiviRemote ID on login'
+    match_on_login_exclude_roles:
```
Comment (dontub) [inline config/schema/civiremote.schema.yml:16]: Should there be a corresponding entry in `civiremote.settings.yml` set to `[]`?

File: src/User.php [hunk 6]
```diff
@@ -45,6 +45,39 @@ public static function create(UserInterface $user) {
     }
   }
 
+  /**
+   * Act on user login.
+   *
+   * @param \Drupal\user\UserInterface $user
+   */
+  public static function login(UserInterface $user) {
```
Comment (dontub) [inline src/User.php]: Return type hint `: void`

**Review by dontub — CHANGES_REQUESTED** [review]

File: config/install/civiremote.settings.yml (hunk 1 above)
Comment (jensschuppe) [reply config/install/civiremote.settings.yml:4]: Current behavior is not to match on login, so no value defaulting to `FALSE` in https://github.com/systopia/civiremote/pull/42/files#diff-4dcbc11d713678e97f70f2321527d825e6e8fc5f6c4f541ee485a050da120f8dR57 should be correct …

**Review by dontub — APPROVED** [review]

**Pull request merged**